Implement tsql_constraintdef for check constraint definition

shalinilohia50 opened the pull request:
This PR contains the following changes:
1. Added **pltsql_ruleutils.c** and **pltsql_ruleutils.h** files: This file contains the rules to de-parse various objects. The functions and structures in this file are taken from PG's ruleutils.c
2. Add **tsql_constraintdef** function to get the check constraint definition
3. Code clean up and basic tests for check constraint definition 
4. Support TSQL typename and typmod de-parsing : Implemented tsql_format_type_extended hook function which returns string representation of type by de-parsing it from type tuples. It will be used
to de-parse data types while de-parsing other object definitions like
constant or variable. Implemented tsql_printTypmod() for printing
typmod in TSQL way.
5. Replace :: with CAST AS function, which is a valid format for TSQL

Task: [BABEL-3222](https://jira.rds.a2z.com/browse/BABEL-3222), BABEL-3146
Authored-by: Harsh Lunagariya <lunharsh@amazon.com>, Shalini Lohia <lshalini@amazon.com>
Signed-off-by: Shalini Lohia <lshalini@amazon.com>

HarshLunagariya committed:
Add pltsql_ruleutils.c and pltsql_ruleutils.h

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

Added tsql_constraintdef for check constraint definition

Fix the build issues to merge it into REL2_X_DEV

Remove few nodetag from isSimpleNode which are not needed for check constraint

Code clean up and basic tests

Signed-off-by: Shalini Lohia <lshalini@amazon.com>

Fix the empty spaces in test output

Signed-off-by: Shalini Lohia <lshalini@amazon.com>

HarshLunagariya committed:
Replace :: with CAST AS function

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

Support TSQL typename and typmod de-parsing

Implemented tsql_format_type_extended hook function which returns string
representation of type by de-parsing it from type tuples. It will be used
to de-parse data types while de-parsing other object definitions like
constant or variable. Implemented tsql_printTypmod() for printing
typmod in TSQL way.

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

Block printing implicit casting in case of operator

Fix 'missing constants for datatypes which has showtype < 0' and dialect
switching

Don't print timezone info and typemod for smalldatetime

Minor fix  in tsql_printTypmod

Remove the hook for tsql_format_type from the engine

Signed-off-by: Shalini Lohia <lshalini@amazon.com>

Remove the file pltsql_format_type

Signed-off-by: Shalini Lohia <lshalini@amazon.com>

shalinilohia50 committed:
Rename the test file, add few more tests and remove the function tsql_format_type_with_typemod

Signed-off-by: SHalini Lohia <lshalini@amazon.com>

Update the expected test output file

shalinilohia50 committed:
Created a view information_schema_tsql.check_constraints,  Added tests covering more nodeTags, cross db and cross schema references

shalinilohia50 committed:
Add order by clause in the tests

HarshLunagariya committed:
Remove implicit casting for relabel node and add functioncall tests

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Minor fix - whitespace removed before AT TIME ZONE

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Replace bbf_binary with binary

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Remove quotes from binary when bbf_binary is encountered

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

rishabhtanwar29 commented on test/JDBC/expected/check_constraints_definition.out:
```diff
@@ -0,0 +1,167 @@
+);
+go
+
+create table test_null(a int, b int, check(a IS NOT NULL), CONSTRAINT constraint1 check (a>10));
```
Can we also add a test for `<> NULL` construct with both `ANSI_NULLS` on and off settings. Something like this:
```
set ansi_nulls to on;
create table test_null1(a int, int b, check(a <> NULL), check(a = NULL);
set ansi_nulls to off;
create table test_null2(a int, int b, check(a <> NULL), check(a = NULL);
```
It is just to verify if we are generating correct constraint def with ansi_nulls settings.

shalinilohia50 replied:
Added a test case for the same.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/pltsql_ruleutils.c:
```diff
@@ -0,0 +1,2441 @@
+		/*
+		 * Replace bbf_binary with binary
+		 */
+		if (strcmp(buf, "bbf_binary") == 0)
```
Don't we also need to replace bbf_varbinary to varbinary?

shalinilohia50 replied:
Added a new column in the test for varbinary, we don't see such a case for varbinary.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/pltsql_ruleutils.c:
```diff
@@ -0,0 +1,2441 @@
+	with_typemod = (flags & FORMAT_TYPE_TYPEMOD_GIVEN) != 0 && (typemod >= 0);
+
+	buf = NULL;
+	LOCAL_FCINFO(fcinfo, 1);
```
Should not mix variable declaration with the implementation. Put `LOCAL_FCINFO` at top with other variable declarations.

shalinilohia50 replied:
Done

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/pltsql_ruleutils.c:
```diff
@@ -0,0 +1,2441 @@
+	*/
+	if (strcmp(typname, "time") == 0 ||
+	   strcmp(typname, "datetime2") == 0 ||
+	   strcmp(typname, "datetimeoffset") == 0)
```
We should also consider the case when typname is schema qualified, in that case the above check won't work.

shalinilohia50 replied:
I have added some tests where the type name was schema qualified.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/pltsql_ruleutils.c:
```diff
@@ -0,0 +1,2441 @@
+	 * AT TIMEZONE from TSQL is parsed to timezone function internally.
+	 * While de-parsing, convert it to AT TIME ZONE explicitly.
+	 */
+	if (strcmp(funcname,"timezone") == 0)
```
Need consider the case when funcname is schema qualified.

shalinilohia50 replied:
Added a test where the function is schema qualified.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/pltsql_ruleutils.c:
```diff
@@ -0,0 +1,2441 @@
+
+	if (showtype < 0)
+	{
+		appendStringInfoString(buf, valbuf->data);
```
We need to free memory of `valbuf->data` before returning

shalinilohia50 replied:
It's addressed in the next commit.

Commit:
Added more test cases

Signed-off-by: Shalini Lohia <lshalini@amazon.com>

HarshLunagariya committed:
Refactor existing datatype checks in type and typemod functions

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

shalinilohia50 committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-3222

shalinilohia50 committed:
Add the ISC view check_constraints to the upgrade script

Signed-off-by: Shalini Lohia <lshalini@amazon.com>

shalinilohia50 committed:
Grant permission to check_constraints ISC view

Deepesh125 commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -17,6 +17,11 @@ RETURNS sys.NVARCHAR(128)
 AS 'babelfishpg_tsql', 'user_name'
 LANGUAGE C IMMUTABLE PARALLEL SAFE;
 
+CREATE OR REPLACE FUNCTION sys.tsql_get_constraintdef(IN login OID DEFAULT NULL)
```
Nitpicks: param name could have been better here. "login" is a bit confusing.

KushaalShroff replied:
I think I had also added the same comments in the previous PR. Looks like they arent addressed yet!

shalinilohia50 replied:
It's been addressed in the next commit.

shalinilohia50 committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-3222

Deepesh125 commented on contrib/babelfishpg_tsql/src/pltsql_ruleutils.c:
```diff
@@ -0,0 +1,2443 @@
+					appendStringInfoChar(buf, '(');
+				get_rule_expr_paren(arg, context, showimplicit, node);
+				appendStringInfo(buf, " COLLATE %s",
+								 generate_collation_name(collate->collOid));
```
We need to be careful here. What would happen if collate->collOid = DEFAULT_COLLATION_OID? would generate_collation_name return "DEFAULT"? if so then is it valid to show it to end user? Probably, it would be okay to show as "DEFAULT" if original constraint was created under PG dialect. But we might need to use Oid corresponding to server_collation_name GUC if original constraint was created under TSQL dialect?

KushaalShroff replied:
> it would be okay to show as "DEFAULT" if original constraint was created under PG dialect
This is work is strictly for TSQL constraints

We should have tests for collate default

shalinilohia50 replied:
Added tests in the next commit.

shalinilohia50 committed:
Better name for the IN parameter name of tsql_get_constraintdef()

Signed-off-by: Shalini Lohia <lshalini@amazon.com>

KushaalShroff commented on contrib/babelfishpg_tsql/sql/information_schema_tsql.sql:
```diff
@@ -401,4 +400,31 @@ CREATE VIEW information_schema_tsql.table_constraints AS
+	SELECT CAST(nc.dbname AS sys.nvarchar(128)) AS "CONSTRAINT_CATALOG",
+		CAST(extc.orig_name AS sys.nvarchar(128)) AS "CONSTRAINT_SCHEMA",
+        	CAST(c.conname AS sys.sysname) AS "CONSTRAINT_NAME",
+		CAST(sys.tsql_get_constraintdef(c.oid) AS sys.nvarchar(4000)) AS "CHECK_CLAUSE"
```
Extra indentation

shalinilohia50 replied:
Fixed it.

KushaalShroff commented on contrib/babelfishpg_tsql/sql/information_schema_tsql.sql:
```diff
@@ -401,4 +400,31 @@ CREATE VIEW information_schema_tsql.table_constraints AS
+		CAST(sys.tsql_get_constraintdef(c.oid) AS sys.nvarchar(4000)) AS "CHECK_CLAUSE"
+
+    FROM sys.pg_namespace_ext nc LEFT OUTER JOIN sys.babelfish_namespace_ext extc ON nc.nspname = extc.nspname,
+         sys.pg_namespace_ext nr LEFT OUTER JOIN sys.babelfish_namespace_ext extr ON nr.nspname = extr.nspname,
```
Please explain why we need nc and nr? Can't we make use of nc only?

shalinilohia50 replied:
Removed the extra join in the next commit.

KushaalShroff commented on test/JDBC/expected/check_constraints_definition.out:
```diff
@@ -0,0 +1,210 @@
+db1#!#dbo#!#test_null2_b_check#!#((b IS NULL))
+db1#!#dbo#!#test_null_a_check#!#((a IS NOT NULL))
+db1#!#dbo#!#test_tsql_cast_c_decimal_check#!#(((c_decimal < CAST((CAST((133.23) AS money)) AS decimal(18,0))) AND (c_decimal < CAST((133.230182309832423) AS decimal(5,3)))))
+db1#!#dbo#!#test_tsql_cast_c_float_check#!#(((c_float < CAST((133.230182309832423) AS float)) AND (c_float < CAST((133.230182309832423) AS real))))
```
We shall document such diffs like float(7) is being displayed as real.

shalinilohia50 replied:
Added this as a comment in the test, will document it in a quip with more differences.

KushaalShroff commented on contrib/babelfishpg_tsql/src/pltsql_ruleutils.h:
```diff
@@ -0,0 +1,12 @@
+/*
+ *
+ * pltsql_ruleutils.h
+ *	  Declarations for pltsql_ruleutils.h
+ *
+ */
+
+#ifndef PLTSQL_RULEUTILS_H
+#define PLTSQL_RULEUTILS_H
+
+#endif							/* PLTSQL_RULEUTILS_H */
+
```
What is the use of this file?

shalinilohia50 replied:
Removed this file in the next commit. It's not needed.

shalinilohia50 committed:
Removed unused file, add test for collate "default"

Removed an extra join from the check_constraints view,
Fixed indentation, removed the unused file pltsql_ruleutils.h,
Added tests for COLLATE "default"

Signed-off-by: Shalini Lohia <lshalini@amazon.com>

rishabhtanwar29 commented on test/JDBC/expected/check_constraints_definition.out:
```diff
@@ -0,0 +1,213 @@
+create table test_null1(a int, b int, check(a <> NULL), check(b = NULL));
+go
+	
+set ansi_nulls off;
```
I think by default ansi_nulls is on so we should reset it to default after this test

rishabhtanwar29 reviewed (approved):
Looks good to me!
